src/Tools/Source/CompilerGeneratorTools/Source/CSharpSyntaxGenerator/CachingSourceGenerator.cs

```diff
@@ -0,0 +1,109 @@
+
+        protected abstract bool TryGenerateSources(
+            AdditionalText input,
+            SourceText inputText,
+            out ImmutableArray<(string hintName, SourceText sourceText)> sources,
+            out ImmutableArray<Diagnostic> diagnostics,
+            CancellationToken cancellationToken);
+
+        public virtual void Initialize(GeneratorInitializationContext context)
+        {
+        }
+
+        public void Execute(GeneratorExecutionContext context)
+        {
+            if (!TryGetRelevantInput(in context, out var input, out var inputText))
+            {
+                return;
+            }
+
+            // Get the current input checksum, which will either be used for verifying the current cache or updating it
+            // with the new results.
+            var currentChecksum = inputText.GetChecksum();
+
+            // Read the current cached result once to avoid race conditions
+            if (s_cachedResult is { } cachedResult
+                && cachedResult.Item1.SequenceEqual(currentChecksum))
+            {
+                AddSources(in context, sources: cachedResult.Item2, currentChecksum, CacheBehavior.None);
+                return;
+            }
+
+            if (TryGenerateSources(input, inputText, out var sources, out var diagnostics, context.CancellationToken))
+            {
+                AddSources(in context, sources, currentChecksum, diagnostics.IsEmpty ? CacheBehavior.Update : CacheBehavior.Clear);
+            }
+            else
+            {
+                Volatile.Write(ref s_cachedResult, null);
+            }
+
+            // Always report the diagnostics (if any)
+            foreach (var diagnostic in diagnostics)
+            {
+                context.ReportDiagnostic(diagnostic);
+            }
+        }
+
+        private static void AddSources(
+            in GeneratorExecutionContext context,
+            ImmutableArray<(string hintName, SourceText sourceText)> sources,
+            ImmutableArray<byte> inputChecksum,
+            CacheBehavior cacheBehavior)
+        {
+            foreach (var (hintName, sourceText) in sources)
+            {
+                context.AddSource(hintName, sourceText);
+            }
+
+            switch (cacheBehavior)
+            {
+                case CacheBehavior.None:
+                    break;
+
+                case CacheBehavior.Clear:
+                    Volatile.Write(ref s_cachedResult, null);
+                    break;
+
+                case CacheBehavior.Update:
+                    // Overwrite the cached result with the new result. This is an opportunistic cache, so as long as
+                    // the write is atomic (which it is for a single pointer) synchronization is unnecessary.
+                    Volatile.Write(ref s_cachedResult, Tuple.Create(inputChecksum, sources));
```

Comment:
Unclear why you are using `Volatile` to read / write this value. It's a `class` so the read / write is atomic. The only value I can see with `Volatile` is forcing the write to be visible sooner but is that really necessary here?

Reply:
It ensures a developer cannot change the `Tuple<>` to `ValueTuple<>` without seeing this comment.

Reply:
Switching to `WeakReference<T>` and a record eliminated the need for this.